src/ts/utils/dom.ts

```diff
@@ -1,41 +1,116 @@
+    
+        let element = document.activeElement;
+        // on some pages like chrome://history/, input is in shadowRoot of several other recursive shadowRoots.
+        while (element?.shadowRoot) {
+            if (element.shadowRoot.activeElement) {
+                element = element.shadowRoot.activeElement;
+            } else {
+                const subElement = element.shadowRoot.querySelector('input, textarea, select');
+                if (subElement) {
+                    element = subElement;
+                }
+                break;
             }
-            break;
         }
-    }
-    return element as ValueElement;
-}
-
-export function getTopLevelBlocks() {
-    return document.querySelector('.roam-article div .flex-v-box') as HTMLElement;
-}
-
-export function getLastTopLevelBlock() {
-  const lastChild = getTopLevelBlocks().lastChild as HTMLElement;
-  return lastChild.querySelector('.roam-block, textarea') as HTMLElement; 
-}
-
-export function getFirstTopLevelBlock() {
-  const firstChild = getTopLevelBlocks().firstChild as HTMLElement;
-  return firstChild.querySelector('.roam-block, textarea') as HTMLElement; 
+        return element as ValueElement;
+    },
+    
+    getTopLevelBlocks() {
+        return document.querySelector('.roam-article div .flex-v-box') as HTMLElement;
+    },
+    
+    getLastTopLevelBlock() {
+      const lastChild = this.getTopLevelBlocks().lastChild as HTMLElement;
+      return lastChild.querySelector('.roam-block, textarea') as HTMLElement; 
+    },
+    
+    getFirstTopLevelBlock() {
+      const firstChild = this.getTopLevelBlocks().firstChild as HTMLElement;
+      return firstChild.querySelector('.roam-block, textarea') as HTMLElement; 
+    },
+    
+    
+    getInputEvent() {
+        return new Event('input', {
+            bubbles: true,
+            cancelable: true,
+        });
+    },
+    
+    async detectChange(fn: () => void, el?: HTMLElement){
+        const targetNode = el || document.querySelector('.roam-body-main') as HTMLElement;
+        const config = { attributes: true, childList: true, subtree: true };
+        return new Promise(async resolve => {
+            //@ts-ignore unused arg
+            const callback = function(mutationsList:MutationRecord[], observer: MutationObserver) {
+                resolve('mutated')
+                observer.disconnect();
+            };
+            const observer = new MutationObserver(callback);
+            observer.observe(targetNode, config);
+            await fn();
+        })
+    },
+    
+    getBlockContainer(element?: HTMLElement) {
+        let parent = null; 
+        let el = element || document.querySelector('textarea') as HTMLElement
+        while(el && !parent){
+            if (el.parentElement?.className?.includes('flex-v-box roam-block-container')) {
+                parent = el?.parentElement;
+            }
+            el = el?.parentElement as HTMLElement;
+        }
```

Comment:
you can query for parent with particular class - see example in my cr https://github.com/roam-unofficial/roam-toolkit/pull/22/files#diff-7f9ad9f2be2cc289f31a201fca94d43cR24